hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneFileSystem.java

```diff
@@ -469,15 +478,72 @@ public boolean delete(Path f, boolean recursive) throws IOException {
         return false;
       }
 
+      OFSPath ofsPath = new OFSPath(key);
+
+      // Handle rm root
+      if (ofsPath.isRoot()) {
+        // Intentionally drop support for rm root
+        // because it is too dangerous and doesn't provide much value
+        LOG.warn("delete: OFS does not support rm root. "
+            + "To wipe the cluster, please re-init OM instead.");
+        return false;
+      }
+
+      // Handle delete volume
+      if (ofsPath.isVolume()) {
+        String volumeName = ofsPath.getVolumeName();
+        if (recursive) {
+          // Delete all buckets first
+          OzoneVolume volume =
+              adapterImpl.getObjectStore().getVolume(volumeName);
+          Iterator<? extends OzoneBucket> it = volume.listBuckets("");
+          while (it.hasNext()) {
+            OzoneBucket bucket = it.next();
+            String nextBucket = addTrailingSlashIfNeeded(
+                f.toString()) + bucket.getName();
+            delete(new Path(nextBucket), true);
+          }
+        }
+        try {
+          adapterImpl.getObjectStore().deleteVolume(volumeName);
+          return true;
+        } catch (OMException ex) {
+          // volume is not empty
+          if (ex.getResult() == VOLUME_NOT_EMPTY) {
+            throw new PathIsNotEmptyDirectoryException(f.toString());
+          } else {
+            throw ex;
+          }
+        }
+      }
+
       result = innerDelete(f, recursive);
+
+      // Handle delete bucket
+      if (ofsPath.isBucket()) {
+        OzoneVolume volume =
+            adapterImpl.getObjectStore().getVolume(ofsPath.getVolumeName());
+        try {
+          volume.deleteBucket(ofsPath.getBucketName());
+          return result;
+        } catch (OMException ex) {
+          // bucket is not empty
+          if (ex.getResult() == BUCKET_NOT_EMPTY) {
+            throw new PathIsNotEmptyDirectoryException(f.toString());
+          } else {
+            throw ex;
+          }
+        }
+      }
+
     } else {
       LOG.debug("delete: Path is a file: {}", f);
       result = adapter.deleteObject(key);
     }
 
     if (result) {
       // If this delete operation removes all files/directories from the
-      // parent direcotry, then an empty parent directory must be created.
+      // parent directory, then an empty parent directory must be created.
```

Comment:
Why is this? Do you mind explaining why do we need to create parent dir. Parent dir should not have deleted right? I know its not related to this change, but just for my understanding.

Reply:
as far as I recall, this should relate to the fact that ozone doesn't create all dirs along the way if we create a key inside a deep directory. e.g. if we create `/volume1/bucket1/dir1/subdir1/key1`, key`/volume1/bucket1/dir1` and `/volume1/bucket1/dir1/subdir1` will NOT be actually created. This breaks some iteration logic (that I couldn't remember exactly which at the moment).

@hanishakoneru should have more info on this.

Reply:
We should remove this logic since HDDS-2940 is in, which solves the problem.

Reply:
some special behavior :-). Its good that we will create dirs along the path now. Thanks for the details.

Reply:
No problem :) Filed HDDS-3596.